### PR TITLE
dev

### PR DESCRIPTION
**Pull request by Alexander8933**
new breanch : dev

### CODE REVIEW

File: src/main/java/com/home/eshop/dao/BikesDao.java [hunk 1]
```diff
@@ -6,15 +6,15 @@
...
+public class BikesDao implements DaoInterface {
     private File file;
     static int maxIdBike = 0;
 
     public BikesDao(String path) {
         this.file = new File(path);
     }
-    private Bike addId(Bike bike) {
+
+    Bike addId(Bike bike) {
```
Comment (sIvanovKonstantyn) [inline src/main/java/com/home/eshop/dao/BikesDao.java:17]: rename to be 'add'

Comment (sIvanovKonstantyn) [reply src/main/java/com/home/eshop/dao/BikesDao.java:17]: or save

File: src/main/java/com/home/eshop/dao/BikesDao.java [hunk 2]
```diff
@@ -26,78 +26,103 @@ private Bike addId(Bike bike) {
...
+            ex.printStackTrace();
         }
         return null;
     }
 
     public List<Bike> findAll() {
-        List<Bike> bikesList = bikes;
-        return bikesList;
+        List<Bike> bikes = new ArrayList<>();
+        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(file))) {
```
Comment (sIvanovKonstantyn) [inline src/main/java/com/home/eshop/dao/BikesDao.java:107]: use stream api

File: src/main/java/com/home/eshop/dao/DaoCache.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public int change(Bike bike) {
+        Optional<Bike> optionalBike = Optional.ofNullable(mapBikes.get(bike.getId()));
+        optionalBike.ifPresent(bikeChang -> {
+            mapBikes.remove(bikeChang.getId());
+            mapBikes.put(bikeChang.getId(), bikeChang);
+        });
+        optionalBike.orElse(bikesDao.findOne(bike.getId()));
```
Comment (sIvanovKonstantyn) [inline src/main/java/com/home/eshop/dao/DaoCache.java:39]: what the logic behind it? explain me in the chat

File: src/main/java/com/home/eshop/dao/BikesDao.java [hunk 4]
```diff
@@ -6,15 +6,15 @@
 import java.util.ArrayList;
 import java.util.List;
 
-public class BikesDao {
-    private ArrayList<Bike> bikes = new ArrayList<Bike>();
+public class BikesDao implements DaoInterface {
```
[the remark this answers is not in the document]
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/BikesDao.java]: Mission Complete

File: src/main/java/com/home/eshop/dao/BikesDao.java (hunk 1 above)
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/BikesDao.java:17]: Rename to idProcessing.
Mission Complete

File: src/main/java/com/home/eshop/dao/BikesDao.java [hunk 5]
```diff
@@ -26,78 +26,103 @@ private Bike addId(Bike bike) {
         return bike;
     }
 
-    public void loadData() {
-        if (bikes.isEmpty()) {
-            try (BufferedReader bufferedReader = new BufferedReader(new FileReader(file))) {
-                String line = null;
-                while ((line = bufferedReader.readLine()) != null) {
-                    addBike(line);
+    public int change(Bike bike) {
```
[the remark this answers is not in the document]
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/BikesDao.java]: Mission Complete

File: src/main/java/com/home/eshop/dao/BikesDao.java [hunk 6]
```diff
@@ -26,78 +26,103 @@ private Bike addId(Bike bike) {
...
+        }
         return bike.getId();
     }
 
     public Bike findOne(int id) {
-        if (id <= maxIdBike) {
-            for (Bike bike : bikes) {
-                if (bike.getId() == id) {
-                    return bikes.get(bikes.indexOf(bike));
+        try (BufferedReader bufferedReader = new BufferedReader(new FileReader(file))) {
```
[the remark this answers is not in the document]
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/BikesDao.java:89]: Mission Complete

File: src/main/java/com/home/eshop/dao/BikesDao.java (hunk 2 above)
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/BikesDao.java:107]: Mission Complete

File: src/main/java/com/home/eshop/dao/DaoCache.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
+package com.home.eshop.dao;
+
+import com.home.eshop.model.Bike;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class DaoCache implements DaoInterface {
```
[the remark this answers is not in the document]
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: Mission Complete

File: src/main/java/com/home/eshop/dao/DaoCache.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    private static void bikeNotFound() {
+        System.out.println("Bike not found");
+    }
+
+    @Override
+    public int change(Bike bike) {
+        Optional<Bike> optionalBike = Optional.ofNullable(mapBikes.get(bike.getId()));
+        optionalBike.ifPresent(bikeChang -> {
+            mapBikes.remove(bikeChang.getId());
```
[the remark this answers is not in the document]
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: Mission Complete

File: src/main/java/com/home/eshop/dao/DaoCache.java (hunk 3 above)
Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: public int update(Bike bike) {
 1   Optional<Bike> optionalBike = Optional.ofNullable(mapBikes.get(bike.getId()));
 2   optionalBike.ifPresent(bikeChang -> {mapBikes.put(bikeChang.getId(), bikeChang);});
 3   optionalBike.orElse(bikesDao.findOne(bike.getId()));
 4   optionalBike.ifPresent(bikeChang -> bikesDao.update(bikeChang));
 5   return bike.getId();}

such logic
1. get (Object or Null) is map ;
2. if  (Object) put the Map ; 
3. if         (Null) get (Object or Null) is txt ;
4. if  (Object) put the txt ;
5. return key ;

old version
 4    optionalBike.ifPresentOrElse(bikeChang -> bikesDao.update(bikeChang), BikesCache::bikeNotFound);
4. if  (Object) put the txt ,if(null) print ("Bike not found")

Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: Mission Complete

Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: Corrected the mistake.

2 optionalBike.ifPresent(bikeChang -> {mapBikes.put(bikeChang.getId(), bike);});
3 optionalBike.orElse(bikesDao.findOne(bike.getId()));
4 optionalBike.ifPresent(bikeChang -> bikesDao.update(bike));

Comment (Alexander8933) [reply src/main/java/com/home/eshop/dao/DaoCache.java]: Mission Complete

File: src/test/java/com/home/eshop/dao/BikesCacheTest.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
+
+class BikesCacheTest {
```
Comment (sIvanovKonstantyn) [inline src/test/java/com/home/eshop/dao/BikesCacheTest.java:13]: Your cache should be tested separately from other pieces of application. Use mockito as was required in a task description.

Comment (Alexander8933) [reply src/test/java/com/home/eshop/dao/BikesCacheTest.java:13]: ok

**Pull request merged**